---
title: "Regression for Microbiome Data:</br>Moving From Diversity to Inference"
subtitle: "ðŸ“šEPID 674ðŸ“š"  
author: "Brendan J. Kelly, MD, MS"
date: 'Updated: `r format(Sys.Date(), "%d %B %Y")`'
output:
  xaringan::moon_reader:
    self_contained: TRUE
    lib_dir: libs
    css: xaringan-themer-plus.css
    nature:
      ratio: 16:9
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---
background-image: url(svg/geompoint.svg)
background-size: 500px
background-position: 85% 50%
class: middle, inverse

```{r setup, include=FALSE}
options(htmltools.dir.version = FALSE)
```

```{r xaringan-themer, include=FALSE, warning=FALSE, eval=FALSE}
library(xaringanthemer)
style_duo_accent(
  primary_color = "#011F5B",
  secondary_color = "#990000",
  inverse_header_color = "#FFFFFF",
  header_font_google = google_font("Roboto Condensed"),
  text_font_google = google_font("Roboto"),
  code_font_google = google_font("Roboto Mono"),
  # padding = "64px 64px 64px 64px"
  # base_font_size = "24px",
  # text_font_base = "1rem",
  # header_h1_font_size = "2.75rem",
  # header_h2_font_size = "2rem",
  # header_h3_font_size = "1.25rem",
)
```

.pad-left[

### Review Î±-diversity

### Review Î²-diversity

### Î±/Î²-diversity â‡¾ linear regression

### Î²-diversity â‡¾ PERMANOVA

### Compositional data?

]



---
background-image: url(svg/geompoint.svg)
background-size: 500px
background-position: 85% 50%
class: center, middle, inverse

# Î±-diversity


---

# High Dimensional Microbiome Data

.center[

```{r otu_table, echo=FALSE, eval=TRUE, warning=FALSE, message=FALSE}

library(tidyverse)

# TYPICAL OTU TABLE ORIENTATION IN MICROBIOME STUDIES

otu <- read_csv("./data/HMP_OTU_table_long.csv.gz")

otu %>%
  reshape2::acast(otu_id ~ specimen_id,
                  # rows = otu_id, columns = specimen_id
                  value.var = "read_count") %>%
     .[1:16,1:6]

# 43140 ROWS & 32 COLUMNS


```

]



---

# High Dimensional Microbiome Data

.pad-left[

- How to deal with high-dimensional microbiome data?

- __Descriptive (e.g., heatmaps and stacked barplots)__
    
- Test a priori hypotheses regarding specific OTUs/taxa

- Reduce dimensions:

    - single summary statistic (alpha diversity)
    
    - pairwise distances (beta diversity) with PCoA or PERMANOVA
    
    - community types (mixture modeling)


]




---
background-image: url(img/hmp_heatmap.png)
background-size: contain



---

# High Dimensional Microbiome Data

.pad-left[

- How to deal with high-dimensional microbiome data?

- Descriptive (e.g., heatmaps and stacked barplots)
    
- Test a priori hypotheses regarding specific OTUs/taxa

- __Reduce dimensions:__

    - __single summary statistic (alpha diversity)__
    
    - pairwise distances (beta diversity) with PCoA or PERMANOVA
    
    - community types (mixture modeling)


]



---

# Shannon Diversity

.pad-left[

- __Richness__ & __evenness__

- Shannon diversity:

    $$H' = - \sum{ p_{i} * \log_{b}{(p_{i})} }$$
    
- "The uncertainty contained in a probability distribution is the average log-probability of an event." (McElreath _Statistical Rethinking, 2nd_ 2020)

]



---
background-image: url(img/hmp_shannon.png)
background-size: contain




---
background-image: url(svg/geompoint.svg)
background-size: 500px
background-position: 85% 50%
class: center, middle, inverse

# Î²-diversity


---

# High Dimensional Microbiome Data

.pad-left[

- How to deal with high-dimensional microbiome data?

- Descriptive (e.g., heatmaps and stacked barplots)
    
- Test a priori hypotheses regarding specific OTUs/taxa

- __Reduce dimensions:__

    - single summary statistic (alpha diversity)
    
    - __pairwise distances (beta diversity) with PCoA or PERMANOVA__
    
    - community types (mixture modeling)


]


---

# Beta Diversity as Dimension Reduction

.pad-left[

- Summarize each sampleâ€™s relationship to other samples:  

    - pairwise distances
    
    - OTU table â†’ square matrix
    
- Many beta diversity metrics:  

    - just counts versus counts + phylogeny
    
    - weighted versus unweighted
    
]



---
background-image: url(img/legendre_otu_to_dm.png)
background-size: contain




---

# Distance Metrics for Beta Diversity

.pad-left[

- Just counts versus counts + phylogeny:

    - Jaccard: $J(A,B) = \frac{Aâˆ©B}{AâˆªB}$  &  $d_{J}(A,B) = 1 - J(A,B)$

    - UniFrac: fraction of unique branch length in tree

- Weighted versus unweighted:

    - weighted: counts matter

    - unweighted: binary (presence-absence)

]



---

# Pairwise Distances â‡¾ PCoA

.pad-left[

- PCoA: principal coordinate analysis

    - any metric distance, even if non-Euclidean
    
    - like PCA, eigenvalue decomposition (maximum variance) but mediated by distance function (no original descriptors)

    - unlike PCA, does not allow projection of original descriptors in reduced-dimension space


]


---
background-image: url(img/weighted_unifrac.png)
background-size: contain



---
background-image: url(img/within_vs_between_group.png)
background-size: contain



---
background-image: url(img/anderson_adonis.png)
background-size: contain



---
background-image: url(img/kelly_bioinformatics_highlight.png)
background-size: contain



---
background-image: url(svg/geompoint.svg)
background-size: 500px
background-position: 85% 50%
class: center, middle, inverse

# Linear regression with Î±/Î²-diversity



---

# Linear Regression with `lm()`

.pull-left[

```{r otu1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

# install.packages("tidyverse") 
library(tidyverse)

# install.packages("vegan")
library(vegan)

# install.packages("ape")
library(ape)

set.seed(16) #<<

otu_tab <- read_rds(
"./data/HMP_OTU_table_matrix_stool_nares.rds"
)

otu_tab %>%
  str(vec.len = 3)



```


]

.pull-right[

```{r otu1-out, ref.label="otu1", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r otu2, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu_tab %>%
  as_tibble(rownames = "otu_id") %>%
  gather(key = "specimen_id",
         value = "read_count",
         -otu_id) %>%
  distinct() -> otu_long

otu_long


```


]

.pull-right[

```{r otu2-out, ref.label="otu2", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r otu3, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

read_tsv(file = "./data/v13_map_uniquebyPSN.txt.bz2") %>%
  rename_all(.funs = ~ gsub("#","",tolower(.x))) %>%
  rename(specimen_id = sampleid) %>%
  distinct() -> specimen_data

specimen_data %>%
  group_by(hmpbodysubsite) %>%
  mutate(dummy_variable_site = #<<
           rnorm(n = length(hmpbodysubsite), #<<
                 mean = nchar(unique(hmpbodysubsite)), #<<
                 sd = 0.5)) %>% #<<
  ungroup() %>%
  filter(hmpbodysubsite %in% #<<
           c("Anterior_nares","Stool")) %>% #<<
  select(specimen_id,
         hmpbodysubsite,
         dummy_variable_site) %>%
  mutate(specimen_id = as.character(specimen_id)) %>%
  distinct() -> specimen_data

specimen_data


```


]

.pull-right[

```{r otu3-out, ref.label="otu3", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r otu4, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

specimen_data %>%
  qplot(data = .,
        x = dummy_variable_site, #<<
        fill = hmpbodysubsite, #<<
        alpha = 0.8,
        geom = "histogram",
        position = "identity") +
  scale_alpha(guide = FALSE) +
  theme_bw() +
  theme(legend.position = "bottom")



```


]

.pull-right[

```{r otu4-out, ref.label="otu4", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r shannon1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu_long %>%
  group_by(specimen_id) %>%
  summarise(shannon = diversity(x = read_count, #<<
                             index = "shannon")) %>% #<<
  ungroup() %>%
  left_join(specimen_data, by = "specimen_id") %>%
  mutate(dummy_variable_shannon = #<<
           rnorm(n = length(shannon), #<<
                 mean = 0, #<<
                 sd = 0.2) + #<<
           shannon) %>% #<<
  distinct() -> shannon_summary

shannon_summary

```


]

.pull-right[

```{r shannon1-out, ref.label="shannon1", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r shannon2, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

shannon_summary %>%
    qplot(data = .,
        x = shannon, #<<
        y = dummy_variable_site, #<<
        color = hmpbodysubsite, #<<
        geom = c("point","smooth"),
        method = "lm") +
  theme_bw() +
  theme(legend.position = "bottom")


```


]

.pull-right[

```{r shannon2-out, ref.label="shannon2", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r shannon2m, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

shannon_summary %>%
  lm(formula = dummy_variable_site ~ shannon, #<<
     data = .) %>%
  summary()


```


]

.pull-right[

```{r shannon2m-out, ref.label="shannon2m", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r shannon3, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

shannon_summary %>%
    qplot(data = .,
        x = shannon, #<<
        y = dummy_variable_shannon, #<<
        color = hmpbodysubsite, #<<
        geom = c("point","smooth"),
        method = "lm") +
  theme_bw() +
  theme(legend.position = "bottom")


```


]

.pull-right[

```{r shannon3-out, ref.label="shannon3", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r shannon3m, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

shannon_summary %>%
  lm(formula = dummy_variable_shannon ~ shannon, #<<
     data = .) %>%
  summary()


```


]

.pull-right[

```{r shannon3m-out, ref.label="shannon3m", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r shannon3mm, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

shannon_summary %>%
  lm(formula = dummy_variable_shannon ~ shannon + #<<
       hmpbodysubsite, #<<
     data = .) %>%
  summary()


```


]

.pull-right[

```{r shannon3mm-out, ref.label="shannon3mm", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r pc1, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu_tab %>%
  t() %>%  # TRANSPOSE #<<
  vegdist(x = ., method = "jaccard") %>% #<<
  pcoa(D = .) -> pc #<<

pc$vectors %>% #<<
  as_tibble(rownames = "specimen_id") %>% #<<
  select(specimen_id, Axis.1, Axis.2) %>% #<<
  left_join(shannon_summary, by = "specimen_id") %>%
  mutate(dummy_variable_pc = 
           rnorm(n = length(shannon), #<<
                 mean = 0, #<<
                 sd = 0.2) + #<<
           Axis.1) %>% #<<
  distinct() -> pc_summary

pc_summary


```


]

.pull-right[

```{r pc1-out, ref.label="pc1", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r pc2, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

pc_summary %>%
  qplot(data = .,
        x = Axis.1, #<<
        y = dummy_variable_pc, #<<
        color = hmpbodysubsite, #<<
        geom = c("point","smooth"),
        method = "lm") +
  theme_bw() +
  theme(legend.position = "bottom")


```


]

.pull-right[

```{r pc2-out, ref.label="pc2", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r pc2m, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

pc_summary %>%
  lm(formula = dummy_variable_pc ~ Axis.1, #<<
     data = .) %>%
  summary()


```


]

.pull-right[

```{r pc2m-out, ref.label="pc2m", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---

# Linear Regression with `lm()`

.pull-left[

```{r pc3m, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

pc_summary %>%
  lm(formula = dummy_variable_pc ~ Axis.2, #<<
     data = .) %>%
  summary()


```


]

.pull-right[

```{r pc3m-out, ref.label="pc3m", echo=FALSE, message=FALSE, warning=FALSE, fig.height=5}

```


]



---
background-image: url(svg/geompoint.svg)
background-size: 500px
background-position: 85% 50%
class: center, middle, inverse

# PERMANOVA with Î²-diversity



---

# PERMANOVA with `adonis()`

.pull-left[

```{r otu11, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

otu_tab %>%
  t() %>%  # TRANSPOSE #<<
  vegdist(x = ., method = "jaccard") -> otu_dist  #<<

otu_dist %>%
  str(vec.len = 2)

```


]

.pull-right[

```{r otu11-out, ref.label="otu11", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# PERMANOVA with `adonis()`

.pull-left[

```{r otu12, echo=TRUE, eval=FALSE, warning=FALSE, message=FALSE}

labels(otu_dist) %>% #match order from dist #<<
  enframe(value = "specimen_id") %>%
  select(specimen_id) %>%
  left_join(pc_summary, by = "specimen_id") %>%
  mutate(dummy_category = Axis.1 > mean(Axis.1)) %>% #<<
  distinct() -> sorted_summary

sorted_summary

```


]

.pull-right[

```{r otu12-out, ref.label="otu12", echo=FALSE, message=FALSE, warning=FALSE}

```


]



---

# PERMANOVA with `adonis()`

.pad-left[

```{r adonis13, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

# distance matrix is response variable #<<
adonis(otu_dist ~ hmpbodysubsite, #<<
       data = sorted_summary) #<<



```


]



---

# PERMANOVA with `adonis()`

.pad-left[

```{r adonis14, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

# multivariable possible... #<<
adonis(otu_dist ~ hmpbodysubsite + dummy_category, #<<
       data = sorted_summary) #<<



```


]



---

# PERMANOVA with `adonis()`

.pad-left[

```{r adonis15, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

# ... but order matters!!! #<<
adonis(otu_dist ~ dummy_category + hmpbodysubsite, #<<
       data = sorted_summary) #<<



```

]


---

# PERMANOVA with `adonis()`

.pad-left[

```{r adonis16, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

# ... do you mean strata? #<<
adonis(otu_dist ~ dummy_category, #<<
       strata = sorted_summary$hmpbodysubsite, #<<
       data = sorted_summary) #<<



```


]




---

# PERMANOVA with `adonis()`

.pad-left[

```{r adonis17, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

# ... or do you mean nestedness? #<<
adonis(otu_dist ~ dummy_category / hmpbodysubsite, #<<
       data = sorted_summary) #<<



```

]


---
background-image: url(svg/geompoint.svg)
background-size: 500px
background-position: 85% 50%
class: center, middle, inverse

# Regression & Compositional Data?



---

# Regression & Compositional Data?

.pad-left[

- Compositional data approaches correct OTU dependency:

    - e.g., `compositions::clr()` or `philr::philr()`

    - p >> n challenges persist

- Must pair compositional transform with regularization:

    - `glmnet::glmnet` for LASSO/ridge/elastic net

    - Bayesian methods

]



---
class: center, middle, inverse
background-image: url(svg/conjugation.svg)
background-size: 500px
background-position: 85% 50%

# Questions?
### Post to the discussion board!


---
background-image: url(svg/bacteria.svg)
background-size: 100px
background-position: 98% 90%
class: center, middle

# Thank you!
#### Slides available: [github.com/bjklab](https://github.com/bjklab/https://github.com/bjklab/EPID674_008_microbiome-regression.git)
#### [brendank@pennmedicine.upenn.edu](brendank@pennmedicine.upenn.edu)




